File: src/main/java/com/cxxcxx/zinecraft/core/registry/ModSound.java
package com.cxxcxx.zinecraft.core.registry;

import com.cxxcxx.zinecraft.api.registry.builder.MusicDiscBuilder;
import com.cxxcxx.zinecraft.core.Zinecraft;

public final class ModSound {
  public static final MusicDiscBuilder AMBIENT_PICTURES_OF_THE_PAST = musicDisc("pictures_of_the_past", 95.0F, "James Primate - Pictures of the Past");
  public static final MusicDiscBuilder AMBIENT_RANDOM_GODS = musicDisc("random_gods", 199.0F, "James Primate - Random Gods (Theme III)");
  public static final MusicDiscBuilder AMBIENT_STRANGER_THINK = musicDisc("stranger_think", 240.0F, "C418 - Stranger Think");
  public static final MusicDiscBuilder AMBIENT_LIFEFLOW = musicDisc("lifeflow", 196.0F, "Monster Siren Records - Lifeflow");
  public static final MusicDiscBuilder AMBIENT_PROCUREMENT_CENTER = musicDisc("procurement_center", 288.67F, "塞壬唱片-MSR - 采购中心");
  public static final MusicDiscBuilder AMBIENT_DABAIXU = musicDisc("dabaixu", 129.38F, "塞壬唱片-MSR - 大柏墟");
  public static final MusicDiscBuilder AMBIENT_JOURNEY_AHEAD = musicDisc("journey_ahead", 202.29F, "塞壬唱片-MSR - 旅途前方");
  public static final MusicDiscBuilder AMBIENT_HUMANITY = musicDisc("humanity", 190.62F, "塞壬唱片-MSR - 人性");
  public static final MusicDiscBuilder AMBIENT_AS_I_SEE_IT = musicDisc("as_i_see_it", 298.48F, "塞壬唱片-MSR - 如我所见");
  public static final MusicDiscBuilder AMBIENT_ESCAPE_PART_2 = musicDisc("escape_part_2", 182.13F, "塞壬唱片-MSR - 逃亡 part2");

  private ModSound() {
  }

  private static MusicDiscBuilder musicDisc(String path, float length, String description) {
    return new MusicDiscBuilder(Zinecraft.SOUNDS, Zinecraft.ITEMS, path, length, description).build();
  }

  public static void bootstrap() {
  }
}

File: core/src/main/kotlin/com/cxxcxx/zinecraft/core/data/ModSound.kt
package com.cxxcxx.zinecraft.core.data

import com.cxxcxx.zinecraft.core.ZinecraftCore
import net.minecraft.core.Holder
import net.minecraft.core.Registry
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.core.registries.Registries
import net.minecraft.data.worldgen.BootstrapContext
import net.minecraft.network.chat.Component
import net.minecraft.resources.ResourceKey
import net.minecraft.sounds.SoundEvent
import net.minecraft.world.item.JukeboxSong

// 参考原版类：SoundEvents
object ModSound {

  val AMBIENT_PICTURES_OF_THE_PAST = Song(
    "ambient.pictures_of_the_past", 95f,
    "James Primate - Pictures of the Past"
  )
  val AMBIENT_RANDOM_GODS = Song(
    "ambient.random_gods", 199f,
    "James Primate - Random Gods (Theme III)"
  )
  val AMBIENT_STRANGER_THINK = Song(
    "ambient.stranger_think", 240f,
    "C418 - Stranger Think"
  )

  fun configure(ctx: BootstrapContext<JukeboxSong>) {
    Song.List.forEach { song -> ctx.register(song.keySong, song.make()) }
  }

  class Song(val name: String, val length: Float, val desc: String) {
    val event: Holder.Reference<SoundEvent> = register(name)
    val keySound: ResourceKey<SoundEvent> = keySound(name)
    val keySong: ResourceKey<JukeboxSong> = keySong(name)
    val signal: Int = 15

    init {
      List.add(this)
    }

    fun make(): JukeboxSong {
      return JukeboxSong(
        event, Component.translatable(name), length, signal
      )
    }

    // 注意，一定要使用 registerForHolder 方法注册
    private fun register(name: String): Holder.Reference<SoundEvent> {
      val r = ZinecraftCore.id(name)
      val o = SoundEvent.createVariableRangeEvent(r)
      return Registry.registerForHolder(BuiltInRegistries.SOUND_EVENT, r, o)
    }

    private fun keySong(name: String): ResourceKey<JukeboxSong> {
      return ZinecraftCore.key(Registries.JUKEBOX_SONG, name)
    }

    private fun keySound(name: String): ResourceKey<SoundEvent> {
      return ZinecraftCore.key(Registries.SOUND_EVENT, name)
    }

    companion object {
      var List: MutableList<Song> = mutableListOf()
    }
  }

  fun init() {}

}